From 09d2f57b4b8709c8f2eefa80262cb47e64d6e4b2 Mon Sep 17 00:00:00 2001
From: Shahed Shaikh <shahedx.shaikh@intel.com>
Date: Mon, 27 Apr 2020 15:02:38 +0800
Subject: [PATCH] DRV_DPM_SW-241: DPM notifier provides garbage subif value

dp_get_port_subitf_via_dev_private() called to get subif info
inside dp_notifier_invoke() fails because of two reasons:
1) if notifier is invoked for ctp_dev, dp_get_port_subitf_via_dev_private()
   does not check sif->ctp_dev and thus it fails to find it's subif info.
2) in deregister_subif flow, DPM sets subif->flags = 0 first
   and then in notifier invocation flow, dp_get_port_subitf_via_dev_private()
   fails because it finds subif->flags = 0.

and dp_notifier_invoke() does not check for failure of
dp_get_port_subitf_via_dev_private() because of this, garbage value
from stack memory allocated for subif_info instance is passed on to
notifier hanlder.

Signed-off-by: Shahed Shaikh <shahedx.shaikh@intel.com>
---
 drivers/net/datapath/dpm/datapath_api.c      | 69 +++++++++++++++++-----------
 drivers/net/datapath/dpm/datapath_notifier.c | 12 +++--
 2 files changed, 52 insertions(+), 29 deletions(-)

diff --git a/drivers/net/datapath/dpm/datapath_api.c b/drivers/net/datapath/dpm/datapath_api.c
index 6ed69de0233d..4984f2d81a10 100644
--- a/drivers/net/datapath/dpm/datapath_api.c
+++ b/drivers/net/datapath/dpm/datapath_api.c
@@ -861,11 +861,6 @@ int32_t dp_deregister_subif_private(int inst, struct module *owner,
 		dp_notifier_invoke(inst, sif->ctp_dev, port_id,
 				   DP_EVENT_DE_REGISTER_SUBIF);
 
-	/* reset mib, flag, and others */
-	memset(&sif->mib, 0, sizeof(sif->mib));
-	sif->flags = 0;
-	sif->netif = NULL;
-	port_info->num_subif--;
 
 	/* subif_hw_reset */
 	if (dp_info->subif_platform_set(inst, port_id, i, &platfrm_data,
@@ -873,6 +868,29 @@ int32_t dp_deregister_subif_private(int inst, struct module *owner,
 		pr_err("subif_platform_set fail\n");
 		/*return res;*/
 	}
+
+	/* for pmapper and non-pmapper both
+	 *  1)for PRX300, dev is managed at its HAL level
+	 *  2)for GRX350, bp/dev should be always zero/NULL at present
+	 *        before adapting to new datapath framework
+	 */
+	if (!get_dp_bp_info(inst, bp)->dev) {
+		DP_DEBUG(DP_DBG_FLAG_REG, "%s for %s inst=%d bp=%d\n",
+			 "dp_inst_del_dev", dev->name, inst, bp);
+
+		dp_inst_del_dev(dev, subif_name, inst, port_id,
+				subif_id->subif, 0);
+
+		dp_notifier_invoke(inst, dev, port_id,
+				   DP_EVENT_DE_REGISTER_SUBIF);
+	}
+
+	/* reset mib, flag, and others */
+	memset(&sif->mib, 0, sizeof(sif->mib));
+	sif->flags = 0;
+	sif->netif = NULL;
+	port_info->num_subif--;
+
 	if (!port_info->num_subif)
 		port_info->status = PORT_DEV_REGISTERED;
 	for (j = 0; j < sif->num_qid; j++) {
@@ -918,21 +936,6 @@ int32_t dp_deregister_subif_private(int inst, struct module *owner,
 			 "cbm_dp_enable ok:port=%d subix=%d cqm_port=%d\n",
 			 port_id, i, cqm_port);
 	}
-	/* for pmapper and non-pmapper both
-	 *  1)for PRX300, dev is managed at its HAL level
-	 *  2)for GRX350, bp/dev should be always zero/NULL at present
-	 *        before adapting to new datapath framework
-	 */
-	if (!get_dp_bp_info(inst, bp)->dev) {
-		DP_DEBUG(DP_DBG_FLAG_REG, "%s for %s inst=%d bp=%d\n",
-			 "dp_inst_del_dev", dev->name, inst, bp);
-
-		dp_inst_del_dev(dev, subif_name, inst, port_id,
-				subif_id->subif, 0);
-
-		dp_notifier_invoke(inst, dev, port_id,
-				   DP_EVENT_DE_REGISTER_SUBIF);
-	}
 
 	DP_DEBUG(DP_DBG_FLAG_REG, "  dp_port=%d subif=%d cqm_port=%d\n",
 		 subif_id->port_id, subif_id->subif, cqm_port);
@@ -1106,11 +1109,6 @@ int32_t dp_update_subif_info(struct dp_subif_upd_info *info)
 					   old_ctp_dev, info->dp_port,
 					   DP_EVENT_DE_REGISTER_SUBIF);
 		}
-		/* added to make the subif_sync info similar to de-register
-		 * subif to force update rcu link list for old_dev/old_ctp_dev
-		 * Later will set this flag correctly
-		 */
-		sif->flags = 0;
 		old_data.ctp_dev = old_ctp_dev;
 	}
 	/* remove old dev */
@@ -1125,6 +1123,12 @@ int32_t dp_update_subif_info(struct dp_subif_upd_info *info)
 				   DP_EVENT_DE_REGISTER_SUBIF);
 		dp_free_bp_priv(info->inst, sif->bp);
 	}
+
+	/* added to make the subif_sync info similar to de-register
+	 * subif to force update rcu link list for old_dev/old_ctp_dev
+	 * Later will set this flag correctly
+	 */
+	sif->flags = 0;
 	/* collect old subif info to update rcu list */
 	old_subif_id_sync = devm_kzalloc(&g_dp_dev->dev,
 					 sizeof(*old_subif_id_sync) * 2,
@@ -1905,6 +1909,15 @@ int dp_get_port_subitf_via_dev_private(struct net_device *dev,
 	int i, j;
 	int inst;
 
+	if (!dev) {
+		pr_err("%s: dev is NULL\n", __func__);
+		return DP_FAILURE;
+	}
+	if (!subif) {
+		pr_err("%s: subif is NULL\n", __func__);
+		return DP_FAILURE;
+	}
+
 	inst = dp_get_inst_via_dev(dev, NULL, 0);
 	for (i = 0; i < get_dp_prop_info(inst)->cap.max_num_dp_ports; i++) {
 		struct pmac_port_info *port = get_dp_port_info(inst, i);
@@ -1914,7 +1927,7 @@ int dp_get_port_subitf_via_dev_private(struct net_device *dev,
 
 			if (!sif->flags)
 				continue;
-			if (sif->netif != dev)
+			if (sif->netif != dev && sif->ctp_dev != dev)
 				continue;
 			subif->port_id = i;
 			subif->subif = SET_VAP(j, port->vap_offset,
@@ -1924,6 +1937,10 @@ int dp_get_port_subitf_via_dev_private(struct net_device *dev,
 			return DP_SUCCESS;
 		}
 	}
+
+	pr_err("%s: subif not found for inst %d dev %s\n", __func__,
+		inst, dev->name);
+
 	return DP_FAILURE;
 }
 
diff --git a/drivers/net/datapath/dpm/datapath_notifier.c b/drivers/net/datapath/dpm/datapath_notifier.c
index 8e6ef634093d..ac5eaf018797 100644
--- a/drivers/net/datapath/dpm/datapath_notifier.c
+++ b/drivers/net/datapath/dpm/datapath_notifier.c
@@ -35,8 +35,9 @@ int dp_notifier_invoke(int inst, struct net_device *dev,
 		       u32 port_id, enum DP_EVENT_TYPE type)
 {
 	struct pmac_port_info *port_info;
-	struct dp_evt_notif_data notif_data;
-	dp_subif_t subif;
+	struct dp_evt_notif_data notif_data = {0};
+	dp_subif_t subif = {0};
+	int ret = DP_SUCCESS;
 
 	port_info = get_dp_port_info(inst, port_id);
 
@@ -57,7 +58,12 @@ int dp_notifier_invoke(int inst, struct net_device *dev,
 		break;
 	case DP_EVENT_REGISTER_SUBIF:
 	case DP_EVENT_DE_REGISTER_SUBIF:
-		dp_get_port_subitf_via_dev_private(dev, &subif);
+		ret = dp_get_netif_subifid_priv(dev, NULL, NULL, NULL, &subif, 0);
+		if (ret) {
+			pr_err("%s: failed to get subif info for %s\n",
+				__func__, dev->name);
+			return DP_FAILURE;
+		}
 		notif_data.subif = subif.subif;
 		break;
 	default:
